crates/cargo-contract/src/cmd/instantiate.rs

```diff
@@ -0,0 +1,309 @@
+    if !instantiate_command.extrinsic_cli_opts.execute {
+        let result = instantiate_exec.instantiate_dry_run().await?;
+        match instantiate_exec.decode_instantiate_dry_run(&result).await {
+            Ok(dry_run_result) => {
+                if instantiate_command.output_json() {
+                    println!("{}", dry_run_result.to_json()?);
+                } else {
+                    print_instantiate_dry_run_result(&dry_run_result);
+                    display_contract_exec_result_debug::<_, DEFAULT_KEY_COL_WIDTH>(
+                        &result,
+                    )?;
+                    display_dry_run_result_warning("instantiate");
+                }
+                Ok(())
+            }
+            Err(object) => {
+                if instantiate_command.output_json() {
+                    return Err(object)
+                } else {
+                    name_value_println!("Result", object, MAX_KEY_COL_WIDTH);
+                    display_contract_exec_result::<_, MAX_KEY_COL_WIDTH>(&result)?;
+                }
+                Err(object)
+            }
+        }
+    } else {
+        tracing::debug!("instantiate data {:?}", instantiate_exec.args().data());
+        let gas_limit = pre_submit_dry_run_gas_estimate_instantiate(
+            &instantiate_exec,
+            instantiate_command.output_json(),
+        )
+        .await?;
+        if !instantiate_command.extrinsic_cli_opts.skip_confirm {
+            prompt_confirm_tx(|| {
+                print_default_instantiate_preview(&instantiate_exec, gas_limit);
+                if let Code::Existing(code_hash) = instantiate_exec.args().code().clone()
+                {
+                    name_value_println!(
+                        "Code hash",
+                        format!("{code_hash:?}"),
+                        DEFAULT_KEY_COL_WIDTH
+                    );
+                }
+            })?;
+        }
+        let instantiate_result = instantiate_exec.instantiate(Some(gas_limit)).await?;
+        display_result(
+            &instantiate_exec,
+            instantiate_result,
+            instantiate_command.output_json(),
+            instantiate_command.extrinsic_cli_opts.verbosity().unwrap(),
+        )
+        .await?;
+        Ok(())
+    }
+}
+
+/// A helper function to estimate the gas required for a contract instantiation.
+async fn pre_submit_dry_run_gas_estimate_instantiate(
+    instantiate_exec: &InstantiateExec,
+    output_json: bool,
+) -> Result<Weight> {
+    if instantiate_exec.opts().skip_dry_run() {
+        return match (instantiate_exec.args().gas_limit(), instantiate_exec.args().proof_size()) {
+                (Some(ref_time), Some(proof_size)) => Ok(Weight::from_parts(ref_time, proof_size)),
+                _ => {
+                    Err(anyhow!(
+                        "Weight args `--gas` and `--proof-size` required if `--skip-dry-run` specified"
+                    ))
+                }
+            };
+    }
```

Comment:
>  If it were an Option<Weight> instead, you can estimate the gas if it's `None`

This logic is already implemented in `call` and `instantiate` methods 
for example

```rust
pub async fn call(
    &self,
    gas_limit: Option<Weight>,
) -> Result<DisplayEvents, ErrorVariant> {
    // use user specified values where provided, otherwise estimate
    let gas_limit = match gas_limit {
        Some(gas_limit) => gas_limit,
        None => self.estimate_gas().await?,
    };
    // ...
    // Logic to call the contract 
```
I think a better approach is to modify `estimate_gas` methods to use what was provided in the builders directly
- if both `gas_limit` and `proof_size` were given, we estimate the gas using them
- otherwise, we do a dry run to estimate the gas

and as I mentioned before, `call` and `instantiate` methods have an optional parameter of type `Weight` so the user of the crate can actually ignore `estimate_gas` completely and pass the maximum gas directly to the methods.

Also, cargo contract crate doesn't use `estimate_gas` at all.
it implements its helper function `pre_submit_dry_run_gas_estimate_instantiate` that does the same logic but with checking for `skip_dry_run` flag and some printing.